src/impl_x86__base_implementation.inl

```diff
@@ -121,7 +121,22 @@ static Leaves ReadLeaves() {
 
 ////////////////////////////////////////////////////////////////////////////////
 // OS support
-// TODO: Add documentation
+// Before an application attempts to use the SIMD extensions,
+// it should check that they are present on the processor and supported by OS:
+//  1. Check that operating system saves and restores XMM/YMM/ZMM/TMM
+//     registers during context switches. Detect CPUID.1:ECX.OSXSAVE[bit 27]
+//     this feature flag specifies that operating system provides extended state
+//     management implied HW support for XSAVE, XRSTOR, XGETBV, XCR0.
+//  2. Check that the SIMD extensions supports by hardware.
+//     Detect CPUID.1:ECX.XSAVE[bit 26]
+//  3. Check enabled state in XCR0 via XGETBV.
+//  4. Check feature flag for instruction set.
+//
+//  Note:
+//  It is unwise for an application to rely exclusively on feature flag for
+//  instruction set or at all on CPUID.1:ECX.XSAVE[bit 26]: These indicate
+//  hardware support but not operating system support. If XMM/YMM/ZMM/TMM state
+//  management is not enabled by an operating systems, instructions will #UD
```

Comment:
AFAIU this is documenting the current behavior right?
If so I would phrase it differently because it reads as if it was an advice to the user.

Also the section to which you added the documentation only contains the following functions:
```
static bool HasMask(uint32_t value, uint32_t mask);
static bool HasXmmOsXSave(uint32_t xcr0_eax);
static bool HasYmmOsXSave(uint32_t xcr0_eax);
static bool HasZmmOsXSave(uint32_t xcr0_eax);
static bool HasTmmOsXSave(uint32_t xcr0_eax);
```

I agree that we could use some documentation to describe what it means to support SIMD instructions but then I would put this section at the top of the file and describe it in terms of process, not in the form of an advice ("it should check", "It is unwise"). Something along these lines:
```
A note on x86 SIMD instructions availability
--------------------------------------------
A number of conditions need to be met for an application to use SIMD instructions:
1. The CPU itself must support the instruction.
    - we use `CPUID` to check whether the feature is supported.
2. The OS must save and restore the associated SIMD register across context switches, we check that:
    - the CPU reports supporting hardware context switching instructions via CPUID.1:ECX.XSAVE[bit 26]
    - the OS reports supporting hardware context switching instructions via CPUID.1:ECX.OSXSAVE[bit 27]
    - the CPU extended control register 0 (XCR0) is set to save and restore the needed SIMD registers

Note that if `XSAVE`/`OSXSAVE` are missing, we delegate the detection to the OS via the `DetectFeaturesFromOs` function or via microarchitecture heuristics.
```